### ISSUE TITLE
Literal Types

### ISSUE THREAD

**Issue by gregbacchus** [pulumi/pulumi-kubernetes#315]
Would it be possible to use Literal Types for things like `apiVersion` and `kind`?

e.g.
```typescript
    export interface Deployment {
      apiVersion?: pulumi.Input<'apps/v1'>
      kind?: pulumi.Input<'Deployment'>
      ...
```

This will help with intellisense and static checking. As it would stop things such as the following being valid:

```typescript
const deployment: apps.v1.Deployment = {
  apiVersion: 'foo',
  kind: 'bar',
  ...
```

**Comment by hausdorff** [comment]
OTOH, `new apps.v1.Deployment` is nice because it's easy to see _where_ the resource registration occurs. This makes the model very declarative, while if it was implicit with something like `const deployment: apps.v1.Deployment = {...}` it's not super clear where/when the resource registration would occur. It would have to be done implicitly somehow when the correct object is built up? So that's a knock against it.

**Comment by CyrusNajmabadi** [comment]
@hausdorff I think the recommendation is *not* to use interfaces over classes.  rather, it is about more tightly typing the types of the *properties*.  i.e. instead of:

```ts
apiVersion?: pulumi.Input<string>;
```

Define things as:

```ts
apiVersion?: pulumi.Input<'apps/v1' | 'whatever' | 'versions' | 'are' | 'supported'>
```

This then gives intellisense/error-checking when you pass an invalid value. This is def possible, and something we do in other repos.

**Comment by lukehoban** [comment]
That is how I interpreted the request as well initially, but on re-reading, I'm actually not sure which of these two things @gregbacchus was interested in (since the example he shared coupled both concepts).  @gregbacchus Any more details on what you were looking for here?

That said, to the extent that there are places where there are enums defined in the OpenAPI spec, it would be nice to project as TS string enums instead of just `string`.

**Comment by gregbacchus** [comment]
I was just interested in the stronger typing. Whether or not that has follow on uses, is probably another discussion.

**Comment by joeduffy** [comment]
This seems like a good thing to do, and not a ton of work given the improvement to typing.

**Comment by hausdorff** [comment]
It's not a ton of work, but Levi is out pretty much until M21, so I'm not sure this will make it.

**Comment by lukehoban** [comment]
@hausdorff Is there strongly typed enum data in the OpenAPI specs that we could use for this?  Or is this about manual annotation?

**Comment by hausdorff** [comment]
@lukehoban Yes, there is, this is already auto-populated in the constructor, we just need to expose it in the typing information too.
